tests/unit/resource/cross_pool_injection.cpp

```diff
@@ -0,0 +1,258 @@
+#include <hpx/lcos/when_all.hpp>
+#include <hpx/runtime/threads/executors/pool_executor.hpp>
+#include <hpx/runtime/threads/threadmanager.hpp>
+#include <hpx/testing.hpp>
+
+#include <cstddef>
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <atomic>
+
+std::size_t num_pools = 0;
+
+// dummy function we will call using async
+void dummy_task(std::size_t n)
+{
+    // no other work can take place on this thread whilst it sleeps
+    std::this_thread::sleep_for(std::chrono::milliseconds(n));
+    //
+    for (std::size_t i(0); i < n; ++i)
+    {
+    }
+}
+
+inline std::size_t st_rand()
+{
+    return std::size_t(std::rand());
+}
+
+int hpx_main(int argc, char* argv[])
+{
+    HPX_TEST_EQ(std::size_t(0), hpx::resource::get_pool_index("default"));
+    HPX_TEST_EQ(std::size_t(0), hpx::resource::get_pool_index("pool-0"));
+
+    // print partition characteristics
+    hpx::threads::get_thread_manager().print_pools(std::cout);
+
+    auto const sched = hpx::threads::get_self_id_data()->get_scheduler_base();
+    if (std::string("core-shared_priority_queue_scheduler") ==
+        sched->get_description())
+    {
+        sched->add_remove_scheduler_mode(
+            // add these flags
+            hpx::threads::policies::scheduler_mode(
+                hpx::threads::policies::enable_stealing |
+                hpx::threads::policies::enable_stealing_numa |
+                hpx::threads::policies::assign_work_thread_parent |
+                hpx::threads::policies::steal_after_local),
+            // remove these flags
+            hpx::threads::policies::scheduler_mode(
+                hpx::threads::policies::assign_work_round_robin |
+                hpx::threads::policies::steal_high_priority_first));
+        sched->update_scheduler_mode(
+            hpx::threads::policies::enable_stealing, false);
+        sched->update_scheduler_mode(
+            hpx::threads::policies::enable_stealing_numa, false);
+    }
+
+    // setup executors for different task priorities on the pools
+    std::vector<hpx::threads::scheduled_executor> HP_executors;
+    std::vector<hpx::threads::scheduled_executor> NP_executors;
+    for (std::size_t i = 0; i < num_pools; ++i)
+    {
+        std::string pool_name = "pool-" + std::to_string(i);
+        HP_executors.push_back(hpx::threads::executors::pool_executor(
+            pool_name, hpx::threads::thread_priority_high));
+        NP_executors.push_back(hpx::threads::executors::pool_executor(
+            pool_name, hpx::threads::thread_priority_default));
+    }
+
+    // randomly create tasks that run on a random pool
+    // attach continuations to them that run on different
+    // random pools
+    const int loops = 10000;
+    //
+    std::cout << "1: Starting HP " << loops << std::endl;
+    std::atomic<int> counter(loops);
+    for (int i = 0; i < loops; ++i)
+    {
+        // high priority
+        std::size_t random_pool_1 = st_rand() % num_pools;
```

Comment:
This could maybe use a uniform distribution from `0` to `num_pools` instead.

Reply:
that would be more 'modern c++' - I'll do it